### PR TITLE
Sparge Attn 

### PR DESCRIPTION
**Pull request by ksikiric**
This pull request introduces support for the Sparge attention backend, including configuration options, backend integration, and corresponding tests. It adds new arguments and configuration fields for Sparge attention, registers new backends, implements backend calls, and ensures proper dependency management and licensing.

The hybrid attention strategy is highly recommended to keep output quality acceptable, the following command:
`xdit     --model Wan2.1-I2V     --ulysses_degree 8    --spargeattn_simthreshold 0.3     --spargeattn_cdfthreshold 0.92     --height 720     --width 1280     --num_frames 81     --num_inference_steps 40     --prompt "Summer beach vacation style, a white cat wearing sunglasses sits on a surfboard. The fluffy-furred feline gazes directly at the camera with a relaxed expression. Blurred beach scenery forms the background featuring crystal-clear waters, distant green hills, and a blue sky dotted with white clouds. The cat assumes a naturally relaxed posture, as if savoring the sea breeze and warm sunlight. A close-up shot highlights the feline's intricate details and the refreshing atmosphere of the seaside." --input_images https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/wan_i2v_input.JPG --cross_attention_backend aiter --use_torch_compile --use_spargeattn_static_block_mask --use_hybrid_attn_schedule     --num_hybrid_attn_high_precision_steps 5     --hybrid_attn_low_precision_backend aiter_sparge     --hybrid_attn_high_precision_backend aiter_sage`

Gives a 9.5% perf gain, compared to running the same config, but with pure AITER SAGE as the attention backend. There is also possible to achieve higher performance gain by being more aggressive (lowering the values) with `spargeattn_simthreshold ` and `spargeattn_cdfthreshold `, but be advised that quality can take a bit hit, finding the correct parameters require some effort. In this PR, some conservative defaults are chosen, balancing performance and output quality. 

https://github.com/user-attachments/assets/e70eccdf-bb7b-473b-80f6-7dad72ad6fd4

Key changes include:

**Sparge Attention Backend Integration:**

* Added two new attention backends: `AITER_SPARGE` and `AITER_SPARGE_V2`, with corresponding implementations in `attention_backend.py` for calling Sparge attention and handling its configuration and block mask logic. 
* Integrated Sparge attention setup, mask computation, and output restoration into the backend, using the new configuration parameters. 
* Registered backend compatibility checks for Sparge attention in `runtime_state.py`, ensuring proper error handling if dependencies are missing or outdated.

**Configuration and CLI Argument Support:**

* Added new fields for Sparge attention configuration (`spargeattn_reorder_sequence`, `use_spargeattn_static_block_mask`, `spargeattn_simthreshold`, `spargeattn_cdfthreshold`) to `RuntimeConfig` and `xFuserArgs`, and propagated them through config creation.
* Extended the CLI argument parser to support these new Sparge attention options, including boolean flags and threshold values.

**Dependency and Licensing Updates:**

* Added `numba` and `numpy` as dependencies in `setup.py` for Sparge attention functionality.
* Updated `LICENSE.txt` to clarify that `xfuser/core/sparge_attention/gilbert.py` is licensed under the BSD 2-Clause License.

### CODE REVIEW

File: xfuser/core/sparge_attention/block_mask.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    B, H, Q, K = cdf.shape
+
+    cdfthreshd_ts = cdfthreshd.view(1, H, 1, 1)
+
+    ge = cdf >= cdfthreshd_ts
+
+    idx = ge.to(torch.uint8).argmax(dim=-1)
+    any_ge = ge.any(dim=-1)
+    # 0-D fallback value broadcasts in `where`; avoids a (B, H, Q) alloc.
+    num_to_select = torch.where(any_ge, idx, idx.new_full((), K))
```
[the remark this answers is not in the document]
Comment (ksikiric) [reply xfuser/core/sparge_attention/block_mask.py:223]: I would say this makes the output quality slightly worse, so I do not trust this comment.

File: xfuser/core/distributed/attention_backend.py [hunk 2]
```diff
@@ -741,9 +754,8 @@ def npu_flash_attn_call(query, key, value, dropout_p, is_causal, attention_kwarg
 @register_attention_function(AttentionBackendType.AITER_SAGE)
 def _aiter_sage_attn_call(query, key, value, dropout_p, is_causal, attention_kwargs=None):
     # Pass layout="bhsd" to avoid permutation
-    softmax_lse = None
-    attn_fn = functools.partial(fav3_sage_wrapper_func, layout="bhsd")
-    output = attn_fn(query, key, value)
+    attn_fn = functools.partial(fav3_sage_wrapper_func, layout="bhsd", return_lse=True)
+    output, softmax_lse = attn_fn(query, key, value)
```
[the remark this answers is not in the document]
Comment (ksikiric) [reply xfuser/core/distributed/attention_backend.py:761]: I will revert the lse for sage, and push a separate PR enabling this and ring attention for Sage

File: xfuser/core/sparge_attention/block_mask.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    if k >= (q + 1) * BqdivBk:
+        tl.store(mask + q * K + k, 0)
+    else:
+        tl.store(mask + q * K + k, 1)
+
+
+def fill_causal_mask_triton(mask: torch.Tensor, BqdivBk:float) -> torch.Tensor:
+    assert mask.dim() == 2
+    triton_fill_causal_mask[mask.shape](mask, BqdivBk)
+    return mask
```
[the remark this answers is not in the document]
Comment (ksikiric) [reply xfuser/core/sparge_attention/block_mask.py:155]: Causal mask is rarely used, so this will not be a high priority to optimize

**Review by feifeibear — APPROVED** [review]
LGTM
